app.py

```diff
@@ -70,6 +71,44 @@
 
 app = Flask(__name__, static_url_path='')
 
+def into_number (string):
+    total = 0
+    for char in string:
+        total += ord (char)
+    return total
+
+def redirect_ab (request, session):
+    # If the user is logged in, we use their username as identifier, otherwise we use the session id
+    user_identifier = current_user(request) ['username'] or str (session_id)
+
+    # This will send 50% of the requests into redirect.
+    redirect_flag = (into_number (user_identifier) % 100) < 50
+    return redirect_flag
+
+# If present, REDIRECT_AB_TEST should be the name of the target environment (that is, environment B).
+if os.getenv ('REDIRECT_AB_TEST') and os.getenv ('REDIRECT_AB_TEST') != os.getenv ('HEROKU_APP_NAME'):
+    @app.before_request
+    def before_request():
+        session_id () # Run this for the side effect of generating a session ID
+        # If we enter this block, we will reverse proxy the request to the REDIRECT_AB_TEST environment.
+        if (redirect_ab (request, session)):
+            url = request.url.replace (os.getenv ('HEROKU_APP_NAME'), os.getenv ('REDIRECT_AB_TEST'))
+
+            request_headers = {}
+            for header in request.headers:
+                if (header [0].lower () in ['host']):
+                    continue
+                request_headers [header [0]] = header [1]
+            r = getattr (requests, request.method.lower ()) (url, headers=request_headers, data=request.data)
```

Comment:
This feels like something Python is not made for, and makes me feel a little uneasy.

Why is making a network request from Python to a second copy of the application preferable to:

* Just having the test behavior INSIDE the one application? (simple)
* If you must, front two different instances of the Flask app with a tool designed to be a reverse proxy, like nginx? (more complex but definitely robust, and imo preferably because the two Flasks are symmetric)

Also, how is the second copy of the application deployed?

Reply:
Okay, I've now understood this was a conscious choice to keep the implementation burden on the second copy of the code base low. 

Doing the proxying in Python still makes me feel a little uneasy. Using nginx seems complicated but I'm going to give it a shot nonetheless.

Reply:
Okay I managed to get it to work in https://github.com/Felienne/hedy/pull/252

I think the solution is mostly solid, but it has as big downside that treatment *cannot*  be based on username (which is what you have now).